### PR TITLE
Task tree: backend performance improvements

### PR DESCRIPTION
**Pull request by psrok1**
This PR is follow-up after https://github.com/CERT-Polska/karton/pull/207

Instead of making extra mappings, we can simply generate task uids in format `{root_uid}:uid`. Most tasks in the Karton environment are "produced" by karton-system while routing, so performance gain will be significant even if most consumers are not upgraded yet to the latest version of Karton.

The only disadvantage is that task identifiers will be a bit lengthy e.g. `{57d3630e-c86d-4931-b071-d1fafb02610d}:f03f9927-dd6a-4275-8706-f922efb7a331`. This may have some impact on the size of logs and Redis memory usage

```
[2024-05-14 12:38:06,884][INFO] Received new task - {57d3630e-c86d-4931-b071-d1fafb02610d}:f03f9927-dd6a-4275-8706-f922efb7a331
{"error": null, "headers": {"origin": "", "quality": "high", "receiver": "karton.wait-for-it", "type": "task"}, "headers_persistent": {"quality": "high"}, "last_update": 1715683086.8855345, "orig_uid": "{57d3630e-c86d-4931-b071-d1fafb02610d}:57d3630e-c86d-4931-b071-d1fafb02610d", "parent_uid": null, "payload": {"payload": {"echo": "off"}}, "payload_persistent": {"__headers_persistent": {"quality": "high"}}, "priority": "normal", "root_uid": "57d3630e-c86d-4931-b071-d1fafb02610d", "status": "Started", "uid": "{57d3630e-c86d-4931-b071-d1fafb02610d}:f03f9927-dd6a-4275-8706-f922efb7a331"}
[2024-05-14 12:38:16,886][INFO] Task done - {57d3630e-c86d-4931-b071-d1fafb02610d}:f03f9927-dd6a-4275-8706-f922efb7a331
[2024-05-14 12:38:16,892][INFO] Received new task - {3d41423a-a99b-43ba-8e89-6c9dcfd71a50}:cbf7c5ec-97da-446d-b65f-1d72997ffada
```

On the other hand, having root_uid in logs is pretty useful to track issues on analysis level and correlate logs with actual samples, so this side-effect is an advantage as well.

To actually get a performance gain from that change, other changes in `karton.inspect` and `karton.backend` have been included as well:

- New KartonBackend method `iter_task_tree` that allows to gather all tasks related with provided `root_uid`. It also lists tasks with legacy uids (matching `karton.task:*[^:]*`) that should be mostly unrouted tasks waiting for `karton-system` processing.
-  `KartonState` is now lazy and gathers information about all tasks in system only when one of `KartonState` properties has been used.  In addition, there is extra method `get_analysis` that allows to get tasks coming only from the specified root_uid.
- `KartonState` has turned off resource parsing by default (`parse_resources=False`). Converting `__karton_resource__` payload entries to actual Resource objects is unnecessary for dashboard and analysis status checking. In the same time, it enables much faster deserialization of tasks.

Changes planned after merging this PR:

- Convert `{root_uid}:uid` back to `uid` in karton-dashboard views to make its views less bloated with long UIDs
- Use `KartonState.get_analysis` in MWDB to make analysis status gathering much quicker.

### CODE REVIEW

File: karton/core/backend.py [hunk 1]
```diff
@@ -534,6 +534,43 @@ def get_all_tasks(
...
+        legacy_task_keys = self.redis.scan_iter(
+            match=f"{KARTON_TASK_NAMESPACE}:*[^:]*", count=chunk_size
+        )
+        for chunk in chunks_iter(legacy_task_keys, chunk_size):
+            yield from filter(
+                lambda task: task.root_uid == root_uid,
+                (
+                    Task.unserialize(task_data, backend=self)
+                    if parse_resources
+                    else Task.unserialize(task_data, parse_resources=False)
```
Comment (msm-cert) [inline karton/core/backend.py:561]: Isn't it just
```suggestion
                    Task.unserialize(task_data, backend=self, parse_resources=parse_resources)
```
?

File: karton/core/backend.py [hunk 2]
```diff
@@ -534,6 +534,43 @@ def get_all_tasks(
...
+        and have the same root_uid
+
+        :param root_uid: Root identifier of task tree
+        :param chunk_size: Size of chunks passed to the Redis SCAN and MGET command
+        :param parse_resources: If set to False, resources are not parsed.
+            It speeds up deserialization. Read :py:meth:`Task.unserialize` documentation
+            to learn more.
+        :return: Iterator with task objects
+        """
+        # Process unrouted tasks produced by Karton <5.4.0
```
Comment (msm-cert) [inline karton/core/backend.py:551]: maybe move this to a separate function if possible (to make it clear it's old functionality, and move it out of the logical "critical path")

File: karton/core/backend.py [hunk 3]
```diff
@@ -534,6 +534,55 @@ def get_all_tasks(
...
+    def _iter_legacy_task_tree(
+        self, root_uid: str, chunk_size: int = 1000, parse_resources: bool = True
+    ) -> Iterator[Task]:
+        """
+        Processes tasks made by <5.4.0 (unrouted from <5.4.0 producers or existing
+        before upgrade)
+
+        Used internally by iter_task_tree.
+        """
+        legacy_task_keys = self.redis.scan_iter(
```
Comment (nazywam) [inline karton/core/backend.py:546]:
```python
 # iterate over all karton tasks that do not match the new task id format
```

File: karton/core/backend.py [hunk 4]
```diff
@@ -534,6 +534,55 @@ def get_all_tasks(
...
+        # Process <5.4.0 tasks (unrouted from <5.4.0 producers
+        # or existing before upgrade)
+        yield from self._iter_legacy_task_tree(
+            root_uid, chunk_size=chunk_size, parse_resources=parse_resources
+        )
+        # Process >=5.4.0 tasks
+        task_keys = self.redis.scan_iter(
+            match=f"{KARTON_TASK_NAMESPACE}:{{{root_uid}}}:*", count=chunk_size
+        )
+        yield from self._iter_tasks(
```
Comment (nazywam) [inline karton/core/backend.py:582]: As a sidenote, I think `_iter_tasks` could be simplified a bit by rewriting it into a standard for loop. List comprehension is cool and all but I think we're overdoing it right now.

File: karton/core/backend.py (hunk 1 above)
Comment (nazywam) [reply karton/core/backend.py]: Not entirely, `backend=self` is not passed for one of the cases.

Comment (psrok1) [reply karton/core/backend.py]: Yeah, but on the other hand there's nothing wrong with not passing the backend parameter when parse_resources=False. It's just not required by the unserialize function in that case

File: karton/core/backend.py (hunk 4 above)
Comment (psrok1) [reply karton/core/backend.py:581]: Simplified in https://github.com/CERT-Polska/karton/pull/257 but good point to avoid such complex expressions

**Review by msm-cert — APPROVED** [review]

**Pull request merged**